### PR TITLE
HAWQ 971

### PR DESCRIPTION
**Pull request by shivzone**
This is based on the earlier pull request for HAWQ 931 https://github.com/apache/incubator-hawq/pull/821
Includes earlier CR fixes and includes updates to return null values for columns not requested in the query

### CODE REVIEW

File: pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/ColumnDescriptor.java [hunk 1]
```diff
@@ -103,12 +112,21 @@ public boolean isKeyColumn() {
...
+
     @Override
 	public String toString() {
 		return "ColumnDescriptor [dbColumnTypeCode=" + dbColumnTypeCode
 				+ ", dbColumnName=" + dbColumnName
 				+ ", dbColumnTypeName=" + dbColumnTypeName
 				+ ", dbColumnIndex=" + dbColumnIndex
-				+ ", dbColumnTypeModifiers=" + dbColumnTypeModifiers + "]";
+				+ ", dbColumnTypeModifiers=" + dbColumnTypeModifiers
+                + ", isProjected=" + isProjected + "]";
```
Comment (sansanichfb) [inline pxf/pxf-api/src/main/java/org/apache/hawq/pxf/api/utilities/ColumnDescriptor.java]: identation

File: src/backend/access/external/pxfheaders.c [hunk 2]
```diff
@@ -165,6 +167,18 @@ static void add_tuple_desc_httpheader(CHURL_HEADERS headers, Relation rel)
         /* Add a key/value pair for attribute type name */
         resetStringInfo(&formatter);
         appendStringInfo(&formatter, "X-GP-ATTR-TYPENAME%u", i);
+
```
Comment (sansanichfb) [inline src/backend/access/external/pxfheaders.c]: This might be not needed anymore since we merged HAWQ-997 and we are sending two new headers X-GP-ATTR-TYPEMODX-COUNT, X-GP-ATTR-TYPEMODX-Y along with all columns which might have modifiers(numeric, varchar, char).

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveInputFormatFragmenter.java [hunk 3]
```diff
@@ -160,6 +166,89 @@ void verifySchema(Table tbl) throws Exception {
...
+     * </ul>
+     * All other types (both in HAWQ and in HIVE) are not supported.
+     * For varchar and char types, typeMod will have the length info along with type.
+     *
+     * @param type HAWQ data type
+     * @param name field name
+     * @return Hive type
+     * @throws UnsupportedTypeException if type is not supported
+     */
+    public static String toHiveType(DataType type, String name, String typeMod) {
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveInputFormatFragmenter.java]: This was moved to HiveUtilities.toCompatibleHiveType when HAWQ-992 has been merged.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java [hunk 4]
```diff
@@ -75,8 +76,6 @@ void parseUserData(InputData input) throws Exception {
         String serdeEnumStr = toks[HiveInputFormatFragmenter.TOK_SERDE];
         if (serdeEnumStr.equals(HiveInputFormatFragmenter.PXF_HIVE_SERDES.COLUMNAR_SERDE.name())) {
             serdeType = HiveInputFormatFragmenter.PXF_HIVE_SERDES.COLUMNAR_SERDE;
-        } else if (serdeEnumStr.equals(HiveInputFormatFragmenter.PXF_HIVE_SERDES.LAZY_BINARY_COLUMNAR_SERDE.name())) {
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java]: Why are we removing support for this serde?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveInputFormatFragmenter.java [hunk 5]
```diff
@@ -160,6 +166,89 @@ void verifySchema(Table tbl) throws Exception {
 
     }
 
+    private void compareTypes(DataType type, String hiveType, String fieldName) {
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveInputFormatFragmenter.java]: This was moved to HiveUtilities.validateTypeCompatible when HAWQ-992 has been merged.

Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveInputFormatFragmenter.java]: Should this function be removed or moved ?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java (hunk 4 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveColumnarSerdeResolver.java]: Accidentally removed

File: src/backend/access/external/pxfheaders.c (hunk 2 above)
Comment (shivzone) [inline src/backend/access/external/pxfheaders.c]: Will remove. I had this update prior to your change

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveInputFormatFragmenter.java (hunk 3 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveInputFormatFragmenter.java]: Yes. Will remove this function as well. Can you make sure the mappings you have configured are the same as here.

File: pxf/pxf-hive/src/test/java/org/apache/hawq/pxf/plugins/hive/HiveDataFragmenterTest.java [hunk 6]
```diff
@@ -52,55 +52,55 @@
...
-        jobConf = mock(JobConf.class);
-        PowerMockito.whenNew(JobConf.class).withArguments(hadoopConfiguration, HiveDataFragmenter.class).thenReturn(jobConf);
-
-        hiveConfiguration = mock(HiveConf.class);
-        PowerMockito.whenNew(HiveConf.class).withNoArguments().thenReturn(hiveConfiguration);
-
-        hiveClient = mock(HiveMetaStoreClient.class);
-        PowerMockito.whenNew(HiveMetaStoreClient.class).withArguments(hiveConfiguration).thenReturn(hiveClient);
-    }
+//    @Test
```
Comment (kavinderd) [inline pxf/pxf-hive/src/test/java/org/apache/hawq/pxf/plugins/hive/HiveDataFragmenterTest.java]: Can you just delete all this commented code?

**Comment by kavinderd** [comment]
Can you remove all references to `Vectorized` Hive classes since they are not yet being used

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+    private void buildArgument(SearchArgument.Builder builder, Object filterObj) {
+        /* The below functions will not be compatible and requires update  with Hive 2.0 APIs */
+        FilterParser.BasicFilter filter = (FilterParser.BasicFilter) filterObj;
+        int filterColumnIndex = filter.getColumn().index();
+        Object filterValue = filter.getConstant().constant();
+        ColumnDescriptor filterColumn = inputData.getColumn(filterColumnIndex);
+        String filterColumnName = filterColumn.columnName();
+
+        switch(filter.getOperation()) {
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: Should we also support HDOP_AND operation? Looks like Builder has method startAnd().

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+import static org.apache.hawq.pxf.api.io.DataType.SMALLINT;
+
+/**
+ * Specialized HiveResolver for a Hive table stored as RC file.
+ * Use together with HiveInputFormatFragmenter/HiveRCFileAccessor.
+ */
+public class HiveORCSerdeResolver extends HiveResolver {
+    private static final Log LOG = LogFactory.getLog(HiveORCSerdeResolver.class);
+    private OrcSerde deserializer;
+    private boolean firstColumn;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java]: Why do we need this variable?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    private StringBuilder builder;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java]: Why do we need this?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+    }
+
+    /*
+     * Get and init the deserializer for the records of this Hive data fragment.
+     * Suppress Warnings added because deserializer.initialize is an abstract function that is deprecated
+     * but its implementations (ColumnarSerDe, LazyBinaryColumnarSerDe) still use the deprecated interface.
+     */
+    @SuppressWarnings("deprecation")
+	@Override
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java]: Identation

File: pxf/pxf-service/src/main/resources/pxf-profiles-default.xml [hunk 11]
```diff
@@ -28,7 +28,8 @@ under the License.
             <plugins>
                 <plugin_A>...</plugin_A>
                 <plugin_B>...</plugin_B>
-                ...
+                ...:q
```
Comment (sansanichfb) [inline pxf/pxf-service/src/main/resources/pxf-profiles-default.xml]: Typo?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/EnumHiveToHawqType.java [hunk 12]
```diff
@@ -152,6 +153,33 @@ public int compare(EnumHiveToHawqType a,
     }
 
     /**
+     *
+     * @return full type name including modifiers. eg: varchar(3)
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/EnumHiveToHawqType.java]: Hive or HAWQ type? I would mention that in JavaDoc and method name.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+import org.apache.hadoop.hive.ql.io.orc.OrcInputFormat;
+import org.apache.hadoop.hive.ql.io.sarg.SearchArgument;
+import org.apache.hadoop.hive.ql.io.sarg.SearchArgumentFactory;
+import org.apache.hadoop.mapred.InputSplit;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: Not used.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hadoop.mapred.Reporter;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: Not used import.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.hawq.pxf.api.FilterParser;
+import org.apache.hawq.pxf.api.utilities.ColumnDescriptor;
+import org.apache.hawq.pxf.api.utilities.InputData;
+import org.apache.commons.lang.StringUtils;
+import java.io.IOException;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: Not used.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        /* Predicate pushdown configuration */
+        String filterStr = inputData.getFilterString();
+        HiveFilterBuilder eval = new HiveFilterBuilder(inputData);
+        Object filter = eval.getFilterObject(filterStr);
+
+        SearchArgument.Builder filterBuilder = SearchArgumentFactory.newBuilder();
+        filterBuilder.startAnd();
+        if (filter instanceof List) {
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: This logic might be moved down to buildArgument.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.commons.lang.CharUtils;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java]: Not used.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.common.type.HiveDecimal;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java]: Not used.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hawq.pxf.api.BadRecordException;
+import org.apache.hawq.pxf.api.OneField;
+import org.apache.hawq.pxf.api.OneRow;
+import org.apache.hawq.pxf.api.UnsupportedTypeException;
+import org.apache.hawq.pxf.api.io.DataType;
+import org.apache.hawq.pxf.api.utilities.ColumnDescriptor;
+import org.apache.hawq.pxf.api.utilities.InputData;
+import org.apache.hawq.pxf.api.utilities.Utilities;
+import org.apache.hawq.pxf.plugins.hdfs.utilities.HdfsUtilities;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java]: Not used.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+import org.apache.hawq.pxf.plugins.hive.utilities.HiveUtilities;
+
+import java.io.IOException;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java]: Not used.

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+import java.math.BigDecimal;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCSerdeResolver.java]: Not used.

File: pxf/pxf-hdfs/src/main/java/org/apache/hawq/pxf/plugins/hdfs/utilities/HdfsUtilities.java [hunk 22]
```diff
@@ -236,6 +236,8 @@ public static Schema getAvroSchema(Configuration conf, String dataSource)
     public static String toString(List<OneField> complexRecord, String delimiter) {
         StringBuilder buff = new StringBuilder();
         String delim = ""; // first iteration has no delimiter
+        if(complexRecord == null)
+            return null;
```
Comment (sansanichfb) [inline pxf/pxf-hdfs/src/main/java/org/apache/hawq/pxf/plugins/hdfs/utilities/HdfsUtilities.java]: Is it safe to return null? Could be better to return empty string?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/EnumHiveToHawqType.java [hunk 23]
```diff
@@ -152,6 +153,33 @@ public int compare(EnumHiveToHawqType a,
...
+     */
+    public static String getFullTypeName(EnumHiveToHawqType hiveToHawqType, Integer[] modifiers) {
+        hiveToHawqType.getTypeName();
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/EnumHiveToHawqType.java]: Why do we call it?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/EnumHiveToHawqType.java [hunk 24]
```diff
@@ -152,6 +153,33 @@ public int compare(EnumHiveToHawqType a,
...
+            for (Integer modifier : modifiers) {
+                if(index++ > 0) {
+                    fullType.append(separator);
+                }
+                fullType.append(modifier);
+            }
+            fullType.append(end);
+            return fullType.toString();
+        } else {
+            return hiveToHawqType.typeName;
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/utilities/EnumHiveToHawqType.java]: Call getter instead of field?

File: pxf/pxf-hive/src/test/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilitiesTest.java [hunk 25]
```diff
@@ -128,47 +128,47 @@ public void mapHiveTypeWithModifiers() throws Exception {
...
-        compatibleTypeName = HiveUtilities.toCompatibleHiveType(DataType.DATE);
+        compatibleTypeName = HiveUtilities.toCompatibleHiveType(DataType.DATE, null);
         assertEquals(compatibleTypeName, EnumHiveToHawqType.DateType.getTypeName());
 
-        compatibleTypeName = HiveUtilities.toCompatibleHiveType(DataType.TIMESTAMP);
+        compatibleTypeName = HiveUtilities.toCompatibleHiveType(DataType.TIMESTAMP, null);
         assertEquals(compatibleTypeName, EnumHiveToHawqType.TimestampType.getTypeName());
 
-        compatibleTypeName = HiveUtilities.toCompatibleHiveType(DataType.NUMERIC);
+        compatibleTypeName = HiveUtilities.toCompatibleHiveType(DataType.NUMERIC, null);
```
Comment (sansanichfb) [inline pxf/pxf-hive/src/test/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilitiesTest.java]: Could you please add tests with not-null modifiers to test getFullTypeName indirectly?

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java (hunk 7 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: On the Java side we are yet to make the update to support both AND and OR logical operators. The cases within buildArgument are strictly only for individual operators.
We can work on the update to support both AND/OR as a separate feature/pull request..

File: pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java (hunk 16 above)
Comment (shivzone) [inline pxf/pxf-hive/src/main/java/org/apache/hawq/pxf/plugins/hive/HiveORCAccessor.java]: I wanted buildArgument to correspond to one search argument and didn't want to make buildArgument a recursive function.

File: pxf/pxf-hive/src/test/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilitiesTest.java (hunk 25 above)
Comment (shivzone) [inline pxf/pxf-hive/src/test/java/org/apache/hawq/pxf/plugins/hive/utilities/HiveUtilitiesTest.java]: Added

**Comment by sansanichfb** [comment]
+1, good job!
